src/Processors/QueryPlan/QueryPlan.cpp

```diff
@@ -543,12 +615,16 @@ void QueryPlan::optimize()
 
         if (frame.next_child == 0)
         {
-            /// First entrance, try push down.
             if (frame.node->children.size() == 1)
             {
                 tryPushDownLimit(frame.node->step, frame.node->children.front());
 
                 while (tryMergeExpressions(frame.node, frame.node->children.front()));
+
+                if (frame.node->children.size() == 1)
+                    tryLiftUpArrayJoin(frame.node, frame.node->children.front(), nodes);
```

Comment:
Currently, it's unclear to me why this tree traversal is correct. Mostly because of changes in tree during traverse:
we add children to the traverse stack, after that change children for some nodes and continue. 
I'm sure that everything works well, but maybe we can clarify this a little bit?

Reply:
Stack contains info for every nodes in the path from tree root to the current node.
Every optimization changes only current node and it's children.
So, optimization does not change any node in stack (except maybe current).
Also, optimization may change query plan step, but not query plan node (only add a new one). So, `node.children` will be always valid.

Also, this code is not the best one. I will try to update it later.